### PR TITLE
Added simple compression/expansion filter

### PR DESCRIPTION
**Pull request by storied**
It probably compiles.

### CODE REVIEW

**Comment by ONSBalder** [comment]
Can one of the admins verify this patch?

File: plugins/obs-filters/compressor-filter.c [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+	float *adata[2] = {(float*)audio->data[0], (float*)audio->data[1]};
+	const size_t channels = cd->channels;
+	const float upper_threshold = cd->upper_threshold;
+	const float lower_threshold = cd->lower_threshold;
+	const float multiplier = cd->multiplier;
+
+	for (size_t i = 0; i < audio->frames; i++) {
+		for (size_t c = 0; c < channels; ++c) {
+			float over = adata[c][i] - upper_threshold;
```
Comment (reboot) [inline plugins/obs-filters/compressor-filter.c]: Is this correct? You create `adata` as an array with a fixed size of 2 elements, but then you interate over it from 0 to (channels - 1), if channels is greater then 2 you will read past the end of the array.

Comment (storied) [inline plugins/obs-filters/compressor-filter.c]: Yeah, if OBS ever supports more than 2 audio channels (I assume it doesn't currently) it's a straight up exploitable stack buffer overflow.

**Comment by hwdro** [comment]
I don't know what I can say about your PR ...
With your default values (upper/lower threshold , multiplier), you offset and attenuate the signal from [-1,1] to [-5.5,-5]. This can't be right.
Also, you're mixing directly decibels (if i understood correctly what upper/lower threshold values mean) with voltage (audio sample), and this isn't right too.
Probably it will be best if you can provide more info about the dsp algorithm used in this "audio filter".

**Comment by storied** [comment]
It should be "fixed" now. For the record, multiplier is just that, a multiplier. Not a ratio, not a gain, a multiplier. No attack and release because most people would use it as a limiter anyway.

**Comment by jp9000** [comment]
I am currently letting other people review this for now.  I won't be able to review this myself properly until 0.12.0, though I was originally just going to implement FFmpeg's compression filter.

Also, forgive me for having to say this, but unless you plan on fixing those commits before an actual merge, there's no way I'd ever merge those commits as-is.  The commits do not abide by the guidelines.

**Comment by storied** [comment]
Not getting merged I guess.

**Comment by jp9000** [comment]
Woa woa, hold on here.  As I stated, I've been busy on 0.12.0 and haven't had time to go through and review it -- I'm already behind on releasing 0.12.0, I've been really trying to get it released ASAP.

**Comment by Metallord** [comment]
A proper dynamic range copression feature would be most welcome. It would make having more professional sound pretty much efortless as hardware solutions are really cumbersome. Even using third party stuff like vsthost is cumbersome. A compression filter right there in obs with sensible defaults gets a :+1: from me. Too bad I'm a filthy casual and don't program C. Anyway this is just some feedback - I really appreciate your time and what you do @jp9000 

P.S. IMHO it shouldn't be "dumbed down" - just give it proper compressor settings with attack, release, etc. Just give it sensible defaults for speech. Really hoping this will come along at some point. Could really help a lot of people sound way better and also save a lot of eardrums :)

**Comment by teddybradford** [comment]
This really would be a fantastic addition to the OBS Studio suite, especially if there was an option to sidechain the filter from, for example, the mic track to the desktop audio track to create a "ducking" effect. I hope someone can take another look at this at some point. :+1:

**Pull request closed without merging**